icu4c/source/common/brkeng.cpp

```diff
@@ -299,6 +306,91 @@ ICULanguageBreakFactory::loadDictionaryMatcherFor(UScriptCode script) {
     return nullptr;
 }
 
+
+URegistryKey ICULanguageBreakFactory::addExternalEngine(
+    ExternalBreakEngine* engine, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return nullptr;
+    }
+    ensureEngines(status);
+    if (U_FAILURE(status)) {
+        return nullptr;
+    }
+    ExternalLanguageBreakEngine* wrapper = new ExternalLanguageBreakEngine(engine);
+    fEngines->push(wrapper, status);
+    return (URegistryKey)wrapper;
+}
+
+UBool ICULanguageBreakFactory::removeExternalEngine(
+    URegistryKey key, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return false;
+    }
+    ensureEngines(status);
+    if (U_FAILURE(status)) {
+        return false;
+    }
+    int32_t found = fEngines->search((ExternalLanguageBreakEngine*)key);
+    if (found < 0) {
+        status = U_ILLEGAL_ARGUMENT_ERROR;
+        return false;
+    }
+    fEngines->removeElementAt(found);
+    return true;
+}
+
+ExternalLanguageBreakEngine::ExternalLanguageBreakEngine(
```

Comment:
Adopting constructors are troublesome for error handling. Ideally you would want to delete  `engine` on failure, but if the memory allocation for the new object fails, the constructor code won't be run, so you can't do it here.

LocalPointers can make it work, although it's a bit obscure. Look up passing std::unique_ptr as function parameters - they're very similar to LocalPointers.
Or, just deal with it at the call site.